py_swf/clients/workflow.py

```diff
@@ -65,3 +70,341 @@ def terminate_workflow(self, workflow_id, reason):
+            oldest_start_date=oldest_start_date,
+            latest_start_date=latest_start_date,
+        )
+        if not start_time_filter_dict:
+            return self.boto_client.count_open_workflow_executions(
+                domain=self.workflow_client_config.domain,
+                **workflow_filter_dict
+            )
+
+        return self.boto_client.count_open_workflow_executions(
+            domain=self.workflow_client_config.domain,
+            startTimeFilter=start_time_filter_dict['startTimeFilter'],
+            **workflow_filter_dict
+        )
+
+    def _count_closed_workflow_executions(
+            self,
+            oldest_start_date=None,
+            latest_start_date=None,
+            oldest_close_date=None,
+            latest_close_date=None,
+            workflow_filter_dict=None
+    ):
+        """
+        Count the number of closed workflows for a domain. You can pass in filtering criteria as a dict
+        The results are best effort and may not exactly reflect recent updates and changes.
+        http://docs.aws.amazon.com/amazonswf/latest/apireference/API_CountClosedWorkflowExecutions.html#SWF-CountClosedWorkflowExecutions-request-executionFilter
+        :param oldest_start_date: datetime
+        :param latest_start_date: datetime
+        :param oldest_close_date: datetime
+        :param latest_close_date: datetime
+        :param workflow_filter_dict: dict. filter type as key and filter criteria as value. filter criteria is a dictionary
+        :return: total number of closed workflows that meet the filter criteria
+        """
+        if not workflow_filter_dict:
+            workflow_filter_dict = {}
+
+        time_filter_dict = _build_time_filter_dict(
+            oldest_start_date=oldest_start_date,
+            latest_start_date=latest_start_date,
+            oldest_close_date=oldest_close_date,
+            latest_close_date=latest_close_date,
+        )
+        if not time_filter_dict:
+            return self.boto_client.count_closed_workflow_executions(
+                domain=self.workflow_client_config.domain,
+                **workflow_filter_dict
+            )
+
+        for key in time_filter_dict:
+            time_filter_type = key
+        if time_filter_type == 'startTimeFilter':
+            return self.boto_client.count_closed_workflow_executions(
+                domain=self.workflow_client_config.domain,
+                startTimeFilter=time_filter_dict[time_filter_type],
+                **workflow_filter_dict
+            )
+        else:
+            return self.boto_client.count_closed_workflow_executions(
+                domain=self.workflow_client_config.domain,
+                closeTimeFilter=time_filter_dict[time_filter_type],
+                **workflow_filter_dict
+            )
+
+
+def _build_time_filter_dict(oldest_start_date=None, latest_start_date=None, oldest_close_date=None, latest_close_date=None):
```

Comment:
Hmm I think the your point on validation part is reasonable. boto client should warn us when calling it wrongly. I will remove the new exception class.
For separating APIs by start date & close date, I think it makes more sense because start_date and close_date are mutually exclusive. A date arg is required for these boto API calls. Separating them makes users easy to understand what date args is required, like a start date is always needed.